File: ts/smelter-core/src/api/output/whip.ts
import type { Api, Outputs, _smelterInternals } from '@swmansion/smelter';
import type { RegisterOutputRequest } from '../output';
import { intoOutputEosCondition, intoVideoEncoderBitrate } from './common';

export function intoRegisterWhipClientOutput(
  output: Outputs.RegisterWhipClientOutput,
  initial: { video?: Api.VideoScene; audio?: Api.AudioScene }
): RegisterOutputRequest {
  return {
    type: 'whip_client',
    endpoint_url: output.endpointUrl,
    bearer_token: output.bearerToken,

    video: intoOutputWhipVideoOptions(output.video, initial.video),
    audio: intoOutputWhipAudioOptions(output.audio, initial.audio),
  };
}

export function intoOutputWhipVideoOptions(
  video: Outputs.WhipVideoOptions | null | undefined,
  initial: Api.VideoScene | undefined
): Api.OutputWhipVideoOptions | undefined {
  if (!video || !initial) {
    return undefined;
  }

  return {
    resolution: video.resolution,
    send_eos_when: video.sendEosWhen && intoOutputEosCondition(video.sendEosWhen),
    encoder_preferences:
      video.encoderPreferences && intoWhipVideoEncoderPreferences(video.encoderPreferences),
    initial,
  };
}

function intoWhipVideoEncoderPreferences(
  encoder_preferences: Outputs.WhipVideoEncoderOptions[]
): Api.WhipVideoEncoderOptions[] {
  return encoder_preferences.map(encoder => {
    switch (encoder.type) {
      case 'ffmpeg_vp9':
        return {
          type: 'ffmpeg_vp9',
          bitrate: encoder.bitrate && intoVideoEncoderBitrate(encoder.bitrate),
          pixel_format: encoder.pixelFormat,
          ffmpeg_options: encoder.ffmpegOptions,
        };
      case 'ffmpeg_vp8':
        return {
          type: 'ffmpeg_vp8',
          bitrate: encoder.bitrate && intoVideoEncoderBitrate(encoder.bitrate),
          ffmpeg_options: encoder.ffmpegOptions,
        };
      case 'ffmpeg_h264':
        return {
          type: 'ffmpeg_h264',
          bitrate: encoder.bitrate && intoVideoEncoderBitrate(encoder.bitrate),
          preset: encoder.preset,
          pixel_format: encoder.pixelFormat,
          ffmpeg_options: encoder.ffmpegOptions,
        };
      case 'vulkan_h264':
        return {
          type: 'vulkan_h264',
          bitrate: encoder.bitrate && intoVideoEncoderBitrate(encoder.bitrate),
        };
      case 'any':
        return {
          type: 'any',
        };
    }
  });
}

function intoOutputWhipAudioOptions(
  audio: true | Outputs.WhipAudioOptions | null | undefined,
  initial: Api.AudioScene | undefined
): Api.OutputWhipAudioOptions | undefined {
  if (!audio || !initial) {
    return undefined;
  }

  if (audio === true) {
    return { initial };
  }

  return {
    send_eos_when: audio.sendEosWhen && intoOutputEosCondition(audio.sendEosWhen),
    channels: audio.channels,
    encoder_preferences:
      audio.encoderPreferences && intoWhipAudioEncoderPreferences(audio.encoderPreferences),
    initial,
  };
}

function intoWhipAudioEncoderPreferences(
  encoder_preferences: Outputs.WhipAudioEncoderOptions[]
): Api.WhipAudioEncoderOptions[] {
  return encoder_preferences.map(encoder => {
    switch (encoder.type) {
      case 'opus':
        return {
          type: 'opus',
          preset: encoder.preset,
          sample_rate: encoder.sampleRate,
          forward_error_correction: encoder.forwardErrorCorrection,
        };
      case 'any':
        return {
          type: 'any',
        };
    }
  });
}

File: compositor_api/src/output/whip.rs
use std::{collections::HashMap, sync::Arc};

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::*;

#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct WhipClient {
    /// WHIP server endpoint
    pub endpoint_url: Arc<str>,
    // Bearer token
    pub bearer_token: Option<Arc<str>>,
    /// Video track configuration.
    pub video: Option<OutputWhipClientVideoOptions>,
    /// Audio track configuration.
    pub audio: Option<OutputWhipClientAudioOptions>,
}

#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct OutputWhipClientVideoOptions {
    /// Output resolution in pixels.
    pub resolution: Resolution,
    /// Defines when output stream should end if some of the input streams are finished. If output includes both audio and video streams, then EOS needs to be sent on both.
    pub send_eos_when: Option<OutputEndCondition>,
    /// Video encoder options.
    pub encoder: Option<VideoEncoderOptions>,
    /// Codec preferences list.
    pub encoder_preferences: Option<Vec<WhipClientVideoEncoderOptions>>,
    /// Root of a component tree/scene that should be rendered for the output.
    pub initial: VideoScene,
}

#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema)]
#[serde(tag = "type", rename_all = "snake_case", deny_unknown_fields)]
pub enum WhipClientVideoEncoderOptions {
    #[serde(rename = "ffmpeg_h264")]
    FfmpegH264 {
        /// (**default=`"fast"`**) Preset for an encoder. See `FFmpeg` [docs](https://trac.ffmpeg.org/wiki/Encode/H.264#Preset) to learn more.
        preset: Option<H264EncoderPreset>,

        /// (**default=`"yuv420p"`**) Encoder pixel format
        pixel_format: Option<PixelFormat>,

        /// Raw FFmpeg encoder options. See [docs](https://ffmpeg.org/ffmpeg-codecs.html) for more.
        ffmpeg_options: Option<HashMap<String, String>>,
    },
    #[serde(rename = "ffmpeg_vp8")]
    FfmpegVp8 {
        /// Raw FFmpeg encoder options. See [docs](https://ffmpeg.org/ffmpeg-codecs.html) for more.
        ffmpeg_options: Option<HashMap<String, String>>,
    },
    #[serde(rename = "ffmpeg_vp9")]
    FfmpegVp9 {
        /// (**default=`"yuv420p"`**) Encoder pixel format
        pixel_format: Option<PixelFormat>,

        /// Raw FFmpeg encoder options. See [docs](https://ffmpeg.org/ffmpeg-codecs.html) for more.
        ffmpeg_options: Option<HashMap<String, String>>,
    },
    #[serde(rename = "any")]
    Any,
}

#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct OutputWhipClientAudioOptions {
    /// (**default="sum_clip"**) Specifies how audio should be mixed.
    pub mixing_strategy: Option<AudioMixingStrategy>,
    /// Condition for termination of output stream based on the input streams states.
    pub send_eos_when: Option<OutputEndCondition>,
    /// Audio encoder options.
    pub encoder: Option<WhipClientAudioEncoderOptions>,
    /// Specifies channels configuration.
    pub channels: Option<AudioChannels>,
    /// Codec preferences list.
    pub encoder_preferences: Option<Vec<WhipClientAudioEncoderOptions>>,
    /// Initial audio mixer configuration for output.
    pub initial: AudioScene,
}

#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema)]
#[serde(tag = "type", rename_all = "snake_case", deny_unknown_fields)]
pub enum WhipClientAudioEncoderOptions {
    Opus {
        /// Specifies channels configuration.
        channels: Option<AudioChannels>,

        /// (**default="voip"**) Specifies preset for audio output encoder.
        preset: Option<OpusEncoderPreset>,

        /// (**default=`48000`**) Sample rate. Allowed values: [8000, 16000, 24000, 48000].
        sample_rate: Option<u32>,

        /// (**default=`false`**) Specifies if forward error correction (FEC) should be used.
        forward_error_correction: Option<bool>,
    },
    Any,
}
